### PR TITLE
feat: TreeData trait

### PR DESCRIPTION
**Pull request by EdJoPaTo**
While a `TreeItem` trait (#33) gives some flexibility and only rendering of existing items a `TreeData` trait allows for fully custom data structures. With a lot of flexibility.

Also, `TreeData` is implemented for a `Vec` of the existing `TreeItem>` so this transition is way more seamless. For JSON this is implemented for `serde_json::Value` directly so it's as simple as `Tree::new(&json)` to display JSON (with the feature enabled).

While it is more complex to implement a good `TreeData` implementation its way less often required as the standard `TreeItem` approach still works and with a lot more performance than before.

tldr: more performance, more flexibility, less required user change than #33.

### CODE REVIEW

File: examples/files.rs [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+use ratatui::widgets::{Block, Scrollbar, ScrollbarOrientation};
+use ratatui::{Frame, Terminal};
+use tui_tree_widget::{Node, Tree, TreeData, TreeState};
+
+struct FileTreeData(PathBuf);
+
+impl TreeData for FileTreeData {
+    type Identifier = OsString;
+
+    fn flatten(&self, open: &HashSet<Vec<Self::Identifier>>) -> Vec<Node<Self::Identifier>> {
```
Comment (joshka) [inline examples/files.rs:20]: Thinking through the ergonomics of some of this (and reading it as someone without actual tree widget experience, perhaps `open_identifiers` would be a better name than `open`, and I wonder if there's a good way of making a specific type or type alias for `Vec<Self::Identifier>` that makes it clear that acts like a path to the identifier - something like `type IdentifierPaths<T> = HashSet<Vec<T>>`, or an actual real type for this? (I'm talking paths to the tree nodes not file paths).

File: examples/files.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            const STYLE: Style = Style::new().fg(Color::DarkGray);
+            spans.push(Span::styled(".", STYLE));
+            spans.push(Span::styled(extension.to_string_lossy(), STYLE));
+        }
+
+        ratatui::widgets::Widget::render(Line::from(spans), area, buffer);
+    }
+}
+
+fn flatten_inner(
```
Comment (joshka) [inline examples/files.rs:53]:
```suggestion
fn flatten_recursive(
```
names this after what it does, rather than where it's used

This could do with a comment about what it does btw for helping those reading the examples find their way.

File: examples/files.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    result: &mut Vec<Node<OsString>>,
+    open: &HashSet<Vec<OsString>>,
+    path: &Path,
+    current: &[OsString],
+) {
+    let Ok(read_dir) = path.read_dir() else {
```
Comment (joshka) [inline examples/files.rs:59]:
```suggestion
    let Ok(entries) = path.read_dir() else {
```

Do you think it could be a good idea to handle the errors by displaying them in the tree instead? E.g.:

```rust
let Ok(entries) = path.read_dir() else {
    result.push(someErrorNode);
    return;
}
for entry in entries {
    let Ok(entry) = entry else {
       result.push(someErrorNode);
       continue;
    }
    ...
}
```

File: examples/files.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+        ratatui::widgets::Widget::render(Line::from(spans), area, buffer);
+    }
+}
+
+fn flatten_inner(
+    result: &mut Vec<Node<OsString>>,
+    open: &HashSet<Vec<OsString>>,
+    path: &Path,
+    current: &[OsString],
```
Comment (joshka) [inline examples/files.rs:57]: Working out what current was took me a bit - it's the current identifier in the tree right?

File: examples/files.rs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    if let Err(err) = res {
+        println!("{err:?}");
+    }
+
+    Ok(())
+}
+
+fn run_app<B: Backend>(terminal: &mut Terminal<B>, mut app: App) -> std::io::Result<()> {
+    const DEBOUNCE: Duration = Duration::from_millis(20); // 50 FPS
```
Comment (joshka) [inline examples/files.rs:150]: Is this necessary for the example? It might over complicate things a bit more than needed.

**Review by joshka — COMMENTED** [review]
It would be helpful to provide some docs that help outline the various types and how they relate to each other as Node, Data, Item, and State all seem very similar conceptually in naming in ways that would be easy to confuse.

File: examples/files.rs (hunk 3 above)
Comment (EdJoPaTo) [reply examples/files.rs:59]: It annoys me more to have no sorting here but I think this should serve as a base example on how to generally approach this.

File: examples/files.rs (hunk 5 above)
Comment (EdJoPaTo) [reply examples/files.rs:150]: It’s the same for all examples and especially without these traits it’s really effective for performance especially on scrolling.

could be moved into a common example base. It seems like too much for this repo so maybe a crate or something behind a feature flag on ratatui? Definitely no perfect solution here.

File: examples/files.rs (hunk 1 above)
Comment (EdJoPaTo) [reply examples/files.rs:20]: I thought about having a dedicated identifier trait instead of `Vec<Identifier>`.

`PathBuf` would allow for everything that is needed without needing the `Vec<OsString>` workaround.
Same with JSON paths that have their own RFC.

File: examples/files.rs (hunk 3 above)
Comment (EdJoPaTo) [reply examples/files.rs:59]: Variable name: the returned struct is named ReadDir which is why I named it that way.
https://doc.rust-lang.org/std/path/struct.Path.html#method.read_dir

File: examples/files.rs (hunk 1 above)
Comment (joshka) [reply examples/files.rs]: Interesting idea. I guess that could work.

File: examples/files.rs (hunk 3 above)
Comment (joshka) [reply examples/files.rs:63]: I mentioned this because not being hugely familiar with the read_dir function, I had to look up what it returns in order to understand the code fully. Naming it entries directly would have made that . `ReadDir` is an `iterator over the entries in a directory.`

**Comment by EdJoPaTo** [comment]
`impl TreeData for Vec<MyTreeItem>` doesnt work for other crates:

```plaintext
error[E0117]: only traits defined in the current crate can be implemented for types defined outside of the crate
…
note: define and implement a trait or new type instead
```

It might be a good idea to create a TreeItem trait and a default `impl TreeData for Vec<TreeItemTrait>` so users can implement the TreeItem trait or TreeData, whichever suits better for them.

**Comment by EdJoPaTo** [comment]
using jsonptr results in worse performance… But its cleaner code and I like the option of using everything and not just `Vec<Something>` so that should stay. For the `jsonptr` stuff I might roll back for performance reasons…

Interestingly, the all_open benchmark is quicker…

```plaintext
render/metadata/no_open time:   [69.148 µs 70.862 µs 72.229 µs]
                        thrpt:  [13.845 Kelem/s 14.112 Kelem/s 14.462 Kelem/s]
                 change:
                        time:   [+4.3049% +11.700% +19.677%] (p = 0.00 < 0.05)
                        thrpt:  [-16.442% -10.475% -4.1273%]
                        Performance has regressed.
render/metadata/few_open
                        time:   [290.15 µs 291.46 µs 292.57 µs]
                        thrpt:  [3.4180 Kelem/s 3.4310 Kelem/s 3.4465 Kelem/s]
                 change:
                        time:   [+22.012% +23.521% +25.152%] (p = 0.00 < 0.05)
                        thrpt:  [-20.097% -19.042% -18.041%]
                        Performance has regressed.
render/metadata/all_open
                        time:   [10.955 ms 11.001 ms 11.049 ms]
                        thrpt:  [90.509  elem/s 90.900  elem/s 91.280  elem/s]
                 change:
                        time:   [-23.692% -23.325% -22.995%] (p = 0.00 < 0.05)
                        thrpt:  [+29.862% +30.421% +31.048%]
                        Performance has improved.
```

**Comment by EdJoPaTo** [comment]
I like seeing https://github.com/fioncat/otree/pull/18 being also somewhat cleaner to implement. @fioncat in case you have feedback and ideas, feel free to join in here :)

Yesterday I added a `GenericTreeItem` trait which simplifies writing your own tree items.
Given that data formats like JSON are key value based I think there should be a specific trait just for them. Currently, I'm not entirely sure on how to do it well as I also don't want extensive cloning of keys. Not doing a trait like that would duplicate a bunch of code in `third_party/json.rs` for other implementations like YAML, TOML, ….

**Comment by EdJoPaTo** [comment]
While the benchmarks result in better performance (not as much as before) the actual FPS in `mqttui` are relatively similar.
On the other hand, using the implementations of `json` / `messagepack` and `TreeData` is way nicer. It's still not perfectly clear how to use the crate ideally. Also, `GenericTreeItem` and `TreeItem` are very close. Thinking about removing `TreeItem` and rename `GenericTreeItem` to `TreeItem` to embrace implementing that trait for end users.

**Comment by EdJoPaTo** [comment]
Tree data structures are surprisingly complex in Rust due to having multiple references of the same thing. This crate shouldn't be about managing these tree structures. It should mainly take them and display them.

But adapting data or even creating some tree data structure in the first place quickly ends up in this pitfall. Which isn't nice.
Managing tree data structures isn't and shouldn't be the core thing of this crate.
When using external crates like `ego_tree` it's all file and possible to handle. So maybe this needs documentation / example pointing users into such direction?

I think the `TreeData` and `GenericTreeItem` traits are great. Stuff like the `ego_tree` in mqttui or data like JSON can be easily displayed. So that part doesn't seem as complex.

**Comment by Teufelchen1** [comment]
As someone who tried using this PR to display some tree/hierarchy: Yes, that's a massive pitfall. Compared to the ease I had with most ratatui widget/data structure this was unexpectedly hard (I did **not** succeed!). While managing such a structure is likely out of scope here, the UX fail is still going to hit this trait/ratatui even if it is not to blame.
